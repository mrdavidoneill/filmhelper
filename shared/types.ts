export interface FilmInfoType {
  Title: string;
  Year: string;
  Rated: string;
  Released: string;
  Runtime: string;
  Genre: string;
  Director: string;
  Writer: string;
  Actors: string;
  Plot: string;
  Language: string;
  Country: string;
  Awards: string;
  Poster: string;
  Ratings: [
    {
      Source: string;
      Value: string;
    },
    {
      Source: string;
      Value: string;
    },
    {
      Source: string;
      Value: string;
    }
  ];
  Metascore: string;
  imdbRating: string;
  imdbVotes: string;
  imdbID: string;
  Type: string;
  DVD: string;
  BoxOffice: string;
  Production: string;
  Website: string;
  Response: "True" | "False";
}

export interface FilmSearchType {
  Title: string;
  Year: string;
  imdbID: string;
  Type: string;
  Poster: string;
}

export interface FilmWatchListType {
  id: number;
  user: string;
  imdb_id: string;
  date_added: string;
}

export interface FilmSearchResponseGoodType {
  Search: FilmSearchType[];
  totalResults: string;
  Response: "True";
}

export interface FilmSearchResponseBadType {
  Error: string;
  Response: "False";
}

export type FilmSearchResponseType =
  | FilmSearchResponseGoodType
  | FilmSearchResponseBadType;

export interface FilmRatingsType {
  id: number;
  rating: number;
  user: string;
  imdb_id: string;
  date_added: string;
}
