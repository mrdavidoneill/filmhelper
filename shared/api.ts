import { FilmInfoType, FilmSearchResponseType } from "@/shared/types";
const API_BASE_URL = process.env.NEXT_PUBLIC_API_BASE_URL;

export async function postAPIRequest({
  body,
  resource,
}: {
  body: Object;
  resource: string;
}) {
  const response = await fetch(`${API_BASE_URL}/${resource}/`, {
    method: "POST",
    headers: {
      "Content-Type": "application/json",
    },
    body: JSON.stringify(body),
  });
  return response.json();
}

export async function getAPIRequest({
  token,
  resource,
}: {
  token: string;
  resource: string;
}) {
  const response = await fetch(`${API_BASE_URL}/${resource}/`, {
    method: "GET",
    headers: {
      "Content-Type": "application/json",
      Authorization: `Token ${token}`,
    },
  });
  return response.json();
}

export async function fetchToken(username: string, password: string) {
  return postAPIRequest({
    body: { username, password },
    resource: "api-token-auth",
  });
}

export async function getFilmWatchList(token: string) {
  return getAPIRequest({
    token,
    resource: "filmwatchlist",
  });
}

export async function searchFilms(
  input: string
): Promise<FilmSearchResponseType> {
  const response = await fetch(
    `http://www.omdbapi.com/?s=${input}&type=movie&apikey=${process.env.NEXT_PUBLIC_OMDB_KEY}`
  );
  return response.json();
}

export async function getFilmInfo(imdbID: string): Promise<FilmInfoType> {
  const response = await fetch(
    `http://www.omdbapi.com/?i=${imdbID}&apikey=${process.env.NEXT_PUBLIC_OMDB_KEY}`
  );
  return response.json();
}
