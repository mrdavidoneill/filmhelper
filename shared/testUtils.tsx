import React, { PropsWithChildren } from "react";
import { render, RenderOptions } from "@testing-library/react";
import { configureStore } from "@reduxjs/toolkit";
import { Provider } from "react-redux";
import type { PreloadedState } from "@reduxjs/toolkit";

import watchListSlice from "@/store/slices/watchListSlice";
import { AppStore, RootState } from "@/store/store";

interface ExtendedRenderOptions extends Omit<RenderOptions, "queries"> {
  preloadedState?: PreloadedState<RootState>;
  store?: AppStore;
}

export function renderWithProviders(
  ui: React.ReactElement,
  {
    preloadedState = {} as RootState,
    // Automatically create a store instance if no store was passed in
    store = configureStore({
      reducer: { watchList: watchListSlice.reducer },
      preloadedState,
    }),
    ...renderOptions
  }: ExtendedRenderOptions = {}
) {
  function Wrapper({ children }: PropsWithChildren<{}>): JSX.Element {
    return <Provider store={store}>{children}</Provider>;
  }

  // Return an object with the store and all of RTL's query functions
  return { store, ...render(ui, { wrapper: Wrapper, ...renderOptions }) };
}
