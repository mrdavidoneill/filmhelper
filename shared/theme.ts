import { blue, deepPurple } from "@mui/material/colors";
import { createTheme } from "@mui/material/styles";
export const theme = createTheme({
  palette: {
    primary: blue,
    secondary: deepPurple,
  },
});
