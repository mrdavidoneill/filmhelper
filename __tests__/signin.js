import "@testing-library/jest-dom";
import React from "react";
import { render, fireEvent } from "@testing-library/react";
import SignIn from "@/pages/signin";
import { ThemeProvider } from "@mui/material";
import { theme } from "@/shared/theme";

jest.mock("@/components/login", () => {
  return function DummyLogin({ setError }) {
    return (
      <button onClick={() => setError("error message")}>Dummy login</button>
    );
  };
});

describe("SignIn", () => {
  it("renders without crashing", () => {
    const { container } = render(
      <ThemeProvider theme={theme}>
        <SignIn />
      </ThemeProvider>
    );

    expect(container).toMatchSnapshot();
  });

  it("shows an error message", () => {
    const { getByText } = render(
      <ThemeProvider theme={theme}>
        <SignIn />
      </ThemeProvider>
    );
    fireEvent.click(getByText("Dummy login"));
    expect(getByText("Error: error message")).toBeInTheDocument();
  });

  it("dismisses the error message", () => {
    const { getByTestId, getByText, queryByText } = render(
      <ThemeProvider theme={theme}>
        <SignIn />
      </ThemeProvider>
    );

    // mock error in login component
    fireEvent.click(getByText("Dummy login"));
    expect(getByText("Error: error message")).toBeInTheDocument();

    // find and click close button
    fireEvent.click(getByTestId("CloseIcon"));

    // assert error message is not displayed
    expect(queryByText("Error: error message")).not.toBeInTheDocument();
  });
});
