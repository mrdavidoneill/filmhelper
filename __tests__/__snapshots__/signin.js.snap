// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SignIn renders without crashing 1`] = `
<div>
  <div
    class="flex flex-col bg-slate-300 h-[100vh] overflow-hidden"
  >
    <button>
      Dummy login
    </button>
  </div>
</div>
`;
